File: scripts/harness_validation/initialization_environment.py
"""初始化验证使用的宿主文件模式辅助函数。"""

from __future__ import annotations

import os
import subprocess
from pathlib import Path

from .context import ROOT


def source_file_has_executable_mode(path: Path) -> bool:
    """按宿主可观测语义确认 Unix 脚本的可执行位。"""

    if os.name != "nt":
        return bool(path.stat().st_mode & 0o111)

    try:
        relative_path = path.resolve().relative_to(ROOT.resolve()).as_posix()  # noqa: F405
    except (OSError, ValueError):
        return False

    result = subprocess.run(
        ["git", "ls-files", "--stage", "--", relative_path],
        cwd=ROOT,  # noqa: F405
        capture_output=True,
        text=True,
        encoding="utf-8",
        errors="replace",
        check=False,
    )
    if result.returncode != 0:
        return False
    entries = [line for line in result.stdout.splitlines() if line.strip()]
    return len(entries) == 1 and entries[0].split(maxsplit=1)[0] == "100755"

File: scripts/harness_validation/initialization_environment.mjs
import fs from "node:fs";

import { run } from "./core.mjs";

/** Check the executable bit stored by Git so Windows and POSIX hosts agree. */
export function sourceFileHasExecutableMode(filePath) {
  if (process.platform !== "win32") {
    try {
      return (fs.statSync(filePath).mode & 0o111) !== 0;
    } catch {
      return false;
    }
  }
  const result = run("git", ["ls-files", "--stage", "--", filePath]);
  if (result.error || result.status !== 0) return false;
  return result.stdout
    .split(/\r?\n/u)
    .filter(Boolean)
    .some((line) => line.startsWith("100755 "));
}
